content/docs/command-reference/exp/remove.md

```diff
@@ -14,9 +14,22 @@ positional arguments:
 
 ## Description
 
-Deletes one or more experiments indicated by name or hash (see `dvc exp run`).
+Deletes one or more experiments, indicated by name (see `dvc exp run`).
 
-With `--queue`, the queue of experiments is cleared.
+With `--queue`, the list of experiments awaiting execution is cleared instead.
+
+> Note that all the checkpoints in an experiment are deleted by this command.
+
+<details>
+
+### What happens to removed experiments?
+
+Internally, `dvc exp` uses actual Git commits under custom Git references (found
+in `.git/refs/exps`). `dvc remove` deletes the references only, rendering the
+commits unreachable. `git` should garbage collect them automatically
+[after some time](https://git-scm.com/docs/git-gc#Documentation/git-gc.txt---pruneltdategt).
```

Comment:
Likewise, I don't think this is something that needs to be documented either. It's a low-level implementation detail.

The DVC docs should just say that removing an experiment works the same way as removing a git branch or tag.

Reply:
You may be right. Then again we have these here and there for the more advanced users. These are in line with the **How does DVC track experiments?** and **How are checkpoints captured by DVC?** sections in [`exp run`](https://dvc.org/doc/command-reference/exp/run).

The reason I decided to include them is that `git`-savvy users may want to manipulate experiments manually or even build integrations (and this is probably hard to figure out from the codebase). WDYT @dberenbaum and/or @shcheklein ? Thanks

p.s. keeping in mind that they are designed as non-intrusive/optional reading:

![image](https://user-images.githubusercontent.com/1477535/112095837-da45de80-8b62-11eb-9ebc-282cdefd6f6d.png)

Reply:
I think it is fine to explain that "experiments are stored as custom git references, and are comparable to branch heads, removing an experiment does the same thing as removing a git branch" and stop there.

But even "git-savvy" users should not be getting into things like manually running `git gc`, or thinking about individual `git` commit objects, etc. Mentioning this kind of things in the documentation is just going to invite user questions along the lines of "why doesn't DVC manually delete commit objects" (the answer being that no one should ever need to manually invoke `git gc` ever, git is smart enough to know when a repo should be garbage collected on its own).

For DVC, just telling the user that the experiment git ref has been deleted should be enough. The underlying details of how git refs/dangling or reachable commits/garbage collection/etc work should not matter to a DVC user. A "git savvy" user can do their own research into how git refs work.

I'd liken this to the github documentation for deleting a branch: https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/managing-branches#deleting-a-branch

Nowhere in their docs does github mention that they create their own duplicate custom refs for all branches, and that when you "delete" a branch on github, it only deletes the standard git branch head ref. The custom ref still exists, so that those commits will never be garbage collected. Github also permanently disables git garbage collection entirely for their server-side repos.

Github does this so that forks always have access to upstream commits which existed at the time the fork was made, and so that users can "restore" previously deleted branches. However, the low-level implementation details don't matter to the user. All the user cares about is whether or not the branch looks "deleted" from the user's UI (and git CLI) perspective. Whether or not those underlying commits still exist is not relevant at all.

A "git savvy" user can use the CLI to inspect all of the refs in the repo and figure out what github is doing, but github does not need to document it.

Reply:
I don't think we can decide what users should be doing or what matters to them, even if we wanted. E.g. that was an argument in https://github.com/iterative/dvc/issues/5181#issuecomment-765261870 yet users seem to care (#5644).

I wonder why `git gc` exists if it's never needed. 🤷 Unrelated

A bit confuse about the quotes around "git savvy" too. Don't those users exist?

> not relevant at all

This kind of statement is hard to understand. Not relevant to whom? Under what criteria?

But anyway, you're probably right that about this being unnecessary detail. Moving that to https://github.com/iterative/dvc.org/pull/2325#pullrequestreview-619177754 (these changes were rolled back in this PR per your feedback).

Reply:
> Not relevant to whom? Under what criteria?

Reading that part again I see this was explained. BTW, very interesting info about Github! Now I know.

Reply:
Gh is a much higher-level tool though, but point taken.

Reply:
`git gc` exists because it may be needed in very specific, advanced use-case situations, if a user is doing low level repo operations like creating git commit/tree/etc objects themselves

> Running git gc manually should only be needed when adding objects to a repository without regularly running such porcelain commands, to do a one-off repository optimization, or e.g. to clean up a suboptimal mass-import.